feat: update expo and refactor things

NeOMakinG opened the pull request:
Sorry, this is kind of a beast to review

## Description
Here is what I did:

- Update expo
- Remove unneeded external dependencies
- Removed unneeded native plugins, use expo to achieve the same thing
- Update both ios and android codes so it works with the new version of expo
- Remove singletons, use useQuery instead (it was crashing expo)
- Use expo env variables over an external dependency
- Leveraging Expo Gyroscope instead of shake lib
- Fix the HMR by leveraging a local variable instead of a state for the gyroscop subscription
- Update all key changes from `app.json`
- Remove metro.config.json, babel.config.json, in favor of expo internals
- Disable deeplinking if the app is running in expo go
- use expo-clipboard (so it works in expo go)


The final result is:

- The app is running in expo go (we don't need an emulator anymore to develop on it)
- The app is running in both ios and android emulator using the build (so we can test the final build or develop using an emulator if you prefer)
- You can use the web shapeshift instance from your local lan if you want to use your phone to develop some things on the shapeshift web app
- We can now configure EAS for: uploading ios and android artifacts from the CI or locally, we can also use OTAs

## How to test
Update your .env file with the proper variables

```
yarn

yarn start
```

Use expo go on your phone to test the app

```
yarn
yarn ios
yarn android
```

You can also try to launch shapeshift with https://github.com/shapeshift/web/pull/9465 , modifying your local url in the env variable (you might need to use `npx expo start -c` to update the env variables cache)

And try the shapeshift web hmr through the expo go app!

So you can test on a local emulator

I didn't really test to build apps using xcode (I lied I did test for this one) and android, I'm not 100% sure it works for now but let's say this will be a second time, depending if we really need considering we are going to use EAS for builds in the future!

NeOMakinG commented on android/app/src/main/AndroidManifest.xml:
```diff
@@ -27,9 +26,8 @@
+        <data android:scheme="shapeshift"/>
       </intent-filter>
     </activity>
-    <activity android:name="com.facebook.react.devsupport.DevSettingsActivity" android:exported="false"/>
```
https://docs.expo.dev/bare/upgrade/?fromSdk=51&toSdk=52

Check the upgrade helpers from 51 to 52 and 52 to 53 for references

NeOMakinG commented on ios/Podfile.lock:
We might need to kick out the Podfile.lock at some point depending on how EAS works (this is required for appflow)

NeOMakinG commented on ios/shapeshift/AppDelegate.swift:
Migrated from Objective-C to Swift because the doc is not including any objective C inputs anymore, so I decided to migrate because we are closer from what expo expose from their documentation (and it was actually fixing the build so... I suspected it's not really compatible with Objectice-C except with maybe some tweaks I don't know about)

NeOMakinG commented on src/App.tsx:
```diff
@@ -35,12 +38,40 @@ const App = () => {
-    const subscription = RNShake.addListener(() => {
-      messageManager.postMessage({ cmd: 'shakeEvent' })
-    })
+    const SHAKE_THRESHOLD = 10.0
+    const SHAKE_TIMEOUT = 250
+    let lastShake = 0
+
+    const handleShake = (data: { x: number; y: number; z: number }) => {
+      const now = Date.now()
+      if (now - lastShake < SHAKE_TIMEOUT) return
+
+      const rotationRate = Math.sqrt(
+        Math.pow(data.x, 2) + Math.pow(data.y, 2) + Math.pow(data.z, 2),
+      )
+
+      if (rotationRate > SHAKE_THRESHOLD) {
+        lastShake = now
+        messageManager.postMessage({ cmd: 'shakeEvent' })
+      }
+    }
+
+    let gyroscopeSubscription: ReturnType<typeof Gyroscope.addListener> | null = null
+
+    ;(async () => {
+      try {
+        await Gyroscope.requestPermissionsAsync()
+        await Gyroscope.setUpdateInterval(100)
+        gyroscopeSubscription = Gyroscope.addListener(handleShake)
+      } catch (gyroscopeError) {
+        console.error('Failed to set up gyroscope:', gyroscopeError)
+      }
+    })()
```
Does approx the same thing as RNShake but with the native gyroscope

NeOMakinG commented on src/App.tsx:
```diff
@@ -126,7 +170,7 @@ const App = () => {
             mediaPlaybackRequiresUserAction={false}
             pullToRefreshEnabled
-            decelerationRate={'normal'}
+            decelerationRate={0.998}
```
React native was crashing with `normal` while using the numeric prop was fine

NeOMakinG commented on src/Root.tsx:
```diff
@@ -1,12 +1,16 @@
+import '@expo/metro-runtime'
```
not 100% sure if we need this tbh, it might be a dust of my hmr fixes tries, happy to remove it if we confirm we don't need it

NeOMakinG replied:
Removed!

NeOMakinG commented on src/components/DeveloperModeModal.tsx:
```diff
@@ -31,68 +17,68 @@ const ENVIRONMENTS: Environment[] = [
     key: 'beard',
     title: 'beard',
-    url: BEARD_URI,
+    url: process.env.EXPO_PUBLIC_BEARD_URI,
```
We can't destructure this thing, expo don't support it, we need to consume them as I did

NeOMakinG commented on src/hooks/useSettings.tsx:
```diff
@@ -1,56 +1,64 @@
+  if (data) {
+    const parsed = JSON.parse(data)
+
+    if (parsed.EXPO_PUBLIC_SHAPESHIFT_URI && isValidUrl(parsed.EXPO_PUBLIC_SHAPESHIFT_URI)) {
```
There is a case where you update the variable, but it's stored in the AsyncStorage, which is peristent between runs, so if you update an env variable, we need to check if it's equal!

gomesalexandre commented on .github/workflows/pr.yml:
```diff
@@ -30,7 +30,7 @@ jobs:
 
       - name: Install
-        run: yarn install --frozen-lockfile
+        run: yarn install --immutable
```
q: you already know what the q is

NeOMakinG replied:
--frozen-lockfile is deprecated, --immutable is more strict and check node_modules also, shouldn't be a big deal for the CI but I did that to remove a warning from the CI

gomesalexandre replied:
Seems sane to me! And web is indeed using that too.

gomesalexandre commented on README.md:
````diff
@@ -67,36 +67,17 @@ Currently we are manually building an archive in xCode to distribute the app.
-```
-5. Run the app on your device from Xcode _or_ use the yarn script:
+### Running on Expo Go with local web
+- Run web, select the lan IP
````
```suggestion
- Run web, copy the lan IP e.g `192.168.x.y`
```

NeOMakinG replied:
Updated in some commits I don't remember

gomesalexandre commented on README.md:
```diff
@@ -67,36 +67,17 @@ Currently we are manually building an archive in xCode to distribute the app.
-- you may need to remove the line `messageManager.webviewRef?.reload()` in the useKeepAliveHook if the device has trouble downloading bundle.js before getting whacked.
-- you will need to manually add your ngrok domain to the `navigationFilter` to prevent opening in a browser. dev flag/mode?
-- as of 10/6: in your local web, move the entire contents of `react-app-rewired/headers/csps` out of their normal place and full restart web (yarn dev)
+- Install Expo Go
```
Probably worth adding Expo go as a prerequisite too, I do have it installed on iOS and know what it is but may not be obvious

NeOMakinG replied:
I completely updated the readme in [4c0cdc4](https://github.com/shapeshift/mobile-app/pull/109/commits/4c0cdc4bcd8120fe53ff1c6b5e319247d80e3d1d)

gomesalexandre replied:
We're now missing Android section, is this expected?

Also probably worth changing the heading of `iOS` explaining what`iOS` (and TBD Android sections) is about - reading this, I would think that I need to refer to `IOS` in addition to `Launching the app` i.e run `yarn ios` if I want to test iOS bits, but reading expo docs, this seems to only be needed for a build and run, we should really only be the case in some very specific cases, is that correct understanding

gomesalexandre replied:
@NeOMakinG still pendarooing

NeOMakinG replied:
You are right, it's needed to launch the build on an simulator, I moved all the ios bits and added the android command in this commit

[13d149d](https://github.com/shapeshift/mobile-app/pull/109/commits/13d149d74f87d32090c24b82a83f6c9622a42f0e)

Most of the time we should be using exop go to develop now that we are compatible with it

gomesalexandre commented on android/app/build.gradle:
I don't think gradle but hey that sounds safe

gomesalexandre commented on android/app/build.gradle:
I don't speak gradle but hey that sounds sane

gomesalexandre commented on android/app/src/main/AndroidManifest.xml:
```diff
@@ -1,6 +1,5 @@
 <manifest xmlns:android="http://schemas.android.com/apk/res/android">
   <uses-permission android:name="android.permission.INTERNET"/>
-  <uses-permission android:name="android.permission.CAMERA"/>
```
q: Is this superseeded by something else i.e does this breaks camera perms?

gomesalexandre replied:
Or perhaps this is the rationale?
https://docs.expo.dev/guides/permissions/#android
> Most permissions are added automatically by libraries that you use in your app either with [config plugins](https://docs.expo.dev/config-plugins/plugins-and-mods/#create-a-plugin) or with a package-level AndroidManifest.xml. You only need to use android.permissions to add additional permissions that are not included by default in a library.

Either way, sounds like it wouldn't hurt to keep explicit?

NeOMakinG replied:
Sent back in [19d6b78](https://github.com/shapeshift/mobile-app/pull/109/commits/19d6b78d18a066a2297b173094cca097e6ae30b8)

Let's see if expo remove it with some codemod at some point in the future

gomesalexandre commented on android/app/src/main/java/com/shapeshift/droid_shapeshift/MainApplication.kt:
```diff
@@ -40,7 +42,7 @@ class MainApplication : Application(), ReactApplication {
   override fun onCreate() {
     super.onCreate()
-    SoLoader.init(this, false)
+    SoLoader.init(this, OpenSourceMergedSoMapping)
```
Ditto LGTM according to migration docs

gomesalexandre commented on android/build.gradle:
```diff
@@ -1,41 +1,37 @@
+  repositories {
+    maven {
+      // All of React Native (JS, Obj-C sources, Android binaries) is installed from npm
+      url(reactNativeAndroidDir)
```
also lgtm

gomesalexandre commented on android/build.gradle:
```diff
@@ -1,41 +1,37 @@
+  }
 }
+
+apply plugin: "expo-root-project"
+apply plugin: "com.facebook.react.rootproject"
```
And ditto! ✅

gomesalexandre commented on ios/Podfile.properties.json:
```diff
@@ -1,4 +1,5 @@
   "expo.jsEngine": "hermes",
-  "EX_DEV_CLIENT_NETWORK_INSPECTOR": "true"
+  "EX_DEV_CLIENT_NETWORK_INSPECTOR": "true",
+  "newArchEnabled": "true"
```
itshappening.jpg

gomesalexandre commented on src/hooks/useKeepAlive.tsx:
```diff
@@ -1,8 +1,7 @@
 import { useEffect, useState } from 'react'
-import { singletonHook } from 'react-singleton-hook'
```
Also used in a single place here so seems sane to me

NeOMakinG replied:
See https://github.com/shapeshift/mobile-app/pull/109#discussion_r2078333976

gomesalexandre commented on src/hooks/useSettings.tsx:
```diff
@@ -1,56 +1,64 @@
-import { useAsyncStorage } from '@react-native-async-storage/async-storage'
-import { useCallback, useEffect, useState } from 'react'
-import { SHAPESHIFT_URI } from 'react-native-dotenv'
-import { singletonHook } from 'react-singleton-hook'
```
ditto only used in a single place, seems sane to me

NeOMakinG replied:
See https://github.com/shapeshift/mobile-app/pull/109#discussion_r2078333976

gomesalexandre commented on package.json:
```diff
@@ -2,52 +2,55 @@
     "react-native-simple-crypto": "latest",
-    "react-native-svg": "15.2.0",
-    "react-native-webview": "13.8.6",
-    "react-singleton-hook": "^4.0.0"
```
After checking all usages, seems like we can indeed safely remove this guy!

gomesalexandre commented on src/hooks/useSettings.tsx:
```diff
@@ -1,56 +1,64 @@
+      return updateSettings(updated)
+    },
+    onSuccess: () => {
+      queryClient.invalidateQueries({ queryKey: [SETTINGS_KEY] })
```
💜

gomesalexandre commented on src/hooks/useSettings.tsx:
```diff
@@ -1,56 +1,64 @@
+    },
+  })
+
+  const setSetting = (name: string, value: unknown) => mutation.mutate({ name, value })
```
Beautiful.

gomesalexandre reviewed (commented):
Oh boi, what a beast. Conceptually didn't see anything standing out, massive work @NeOMakinG 🐐 

Very very extensive runtime pass to follow

gomesalexandre commented on README.md:
```diff
@@ -67,36 +67,17 @@ Currently we are manually building an archive in xCode to distribute the app.
-- as of 10/6: in your local web, move the entire contents of `react-app-rewired/headers/csps` out of their normal place and full restart web (yarn dev)
+- Install Expo Go
+- Launch ShapeShift by scanning the QR code
+- Select the `local` environment in the settings
```
```suggestion
- Select the `localhost` environment in the settings
```

NeOMakinG replied:
Changed in [19d6b78](https://github.com/shapeshift/mobile-app/pull/109/commits/19d6b78d18a066a2297b173094cca097e6ae30b8)

gomesalexandre reviewed (commented):
Tested locally:

- Wallet creation is happy ✅ 
- Wallet import is happy and fetches accounts ✅ 
- Switching to localhost env is happy ✅ 
- Trade: Unable to continue, failed to estimate network fees, may or may not be related to my current network shenanigans ❓ 
- Onramper is sad, this does *not* seem related to the above, probably the `crypto` issue previously mentioned 🚫 
- QR code scanning is sad
![IMG_3363](https://github.com/user-attachments/assets/1920953f-f71b-410f-bb57-2faaa050a963)
- Copy is sad 🚫 
![IMG_3359](https://github.com/user-attachments/assets/50d8641e-96f1-4693-beda-03336a88ef02)

@NeOMakinG happy to retest these after they are fixed, though because of my current network shenanigans, I won't be able to fully e2e this either way

NeOMakinG commented on src/hooks/useImportWallet.tsx:
```diff
@@ -1,9 +1,11 @@
+import AsyncStorage from '@react-native-async-storage/async-storage'
 import { deleteItemAsync, getItemAsync } from 'expo-secure-store'
-import { useEffect, useState } from 'react'
-import { singletonHook } from 'react-singleton-hook'
```
It was completely crashing because of infinit state loops without it, but it was more of a weird anti-pattern patch than an actual need, I removed it in favor of useQuery so now these hooks are properly reusable and not causing unnecessary rerenders anymore!

NeOMakinG commented on package.json:
```diff
@@ -2,52 +2,55 @@
     "@types/lodash.once": "^4.1.7",
-    "@types/react": "~18.2.79",
-    "@types/react-native": "^0.73.0",
+    "@types/react": "~19.0.10",
```
Done in [19d6b78](https://github.com/shapeshift/mobile-app/pull/109/commits/19d6b78d18a066a2297b173094cca097e6ae30b8)

NeOMakinG commented on src/App.tsx:
```diff
@@ -59,15 +90,18 @@ const App = () => {
       setUri(newUri)
     }
 
+    if (isRunningInExpoGo) return
```
I removed this by changed the url parsing to `exp://` if it's running inside expo go!

So now we does support deeplinks through expo go anyway and so we don't need to early return

gomesalexandre replied:
As discussed, doesn't hurt to keep even though we don't use this atm but may, meh

NeOMakinG replied:
Updated comments in [741d88c](https://github.com/shapeshift/mobile-app/pull/109/commits/741d88cd200caa6d9717cb184a525b018ae0c4e6)

Removed the yat bits as we don't support them anymore anyway

Keeping this bits for the sake of supporting deeplinks if ever someone did use it or will use in the future (aka swapper redirect prefill maybe)

NeOMakinG replied:
Step back on this: [d012fa2](https://github.com/shapeshift/mobile-app/pull/109/commits/d012fa2cb3410b007beedb133e6db47bd23a54ea)

I early returned (+ comment) on expo go as we obviously can't support deeplink through it as Expo Go is an app by itself, it will even send unwanted event we can't handle

NeOMakinG commented on src/App.tsx:
```diff
@@ -35,12 +38,40 @@ const App = () => {
-    const subscription = RNShake.addListener(() => {
-      messageManager.postMessage({ cmd: 'shakeEvent' })
-    })
+    const SHAKE_THRESHOLD = 10.0
```
Removed it in [19d6b78](https://github.com/shapeshift/mobile-app/pull/109/commits/19d6b78d18a066a2297b173094cca097e6ae30b8)

NeOMakinG commented on android/gradle.properties:
```diff
@@ -54,3 +56,7 @@ EX_DEV_CLIENT_NETWORK_INSPECTOR=true
 expo.useLegacyPackaging=false
+
+
+# Whether the app is configured to use edge-to-edge via the app config or `react-native-edge-to-edge` plugin
+expo.edgeToEdgeEnabled=false
```
I think Expo updated it or some kind of copy pasta from a raw configuration

NeOMakinG commented:
So, I had a workaround for the copy borked, fixed it

I've been able to send swaps, seems related to your connection

Other parts: Camera and onRamper, are working using one of our environment or localhost, but will never work on the local IP as far it's an insecure environment unfortunately, I couldn't find a solution yet and I don't even know if we will find one at some point or if it's worth the effort but:

If we are touching parts requiring an secure environment, we could use the emulator as previously, with `http://localhost:3000`, this is the workaround we have for now!

gomesalexandre commented on package.json:
```diff
@@ -2,52 +2,54 @@
+  "main": "index.js",
   "scripts": {
     "clean": "react-native-clean",
     "clean:ios": "rm -rf ios/build && rm -rf ios/Pods && react-native-clean",
```
Also this isn't mentioned anywhere in the README, either remove me or make use of it there 🙏🏽

NeOMakinG replied:
Sorry I did miss this comment, see https://github.com/shapeshift/mobile-app/pull/109#issuecomment-2866659306

I removed this in favor of `expo prebuild` which should be saner considering the current setup we have now

I would love you to give it a try in case you have a borked env, as my env is working fine (damn mobile)

gomesalexandre reviewed (changes requested):
Retested locally

### Expo go (iOS)

- clippy is happy 🚫 

While it *seems* to be happy (no error), it doesn't actually copy. That's using expo go with the regular prod env

- Swaps are happy ❓ 

- Not sure if this is an issue with the app itself or expo go, but switching apps and coming back hot reloads back from splash screen, which doesn't seem expected at all, and make things like testing creating a temp wallet for testing (without writing down the words and just taking a temp screenshot) virtually impossible

- CoW swaps (message signing, no approval) are happy ✅ 

- Actual Tx broadcast swaps are happy ✅ : somehow happy for me now, not sure what was happening previously! Tested with Tx approval and broadcast 

- Home assets page is happy ✅ 

- Non-Onramper (e.g Banxa) fiat ramps are happy ✅ 

- FOX Page is happy ✅ 

- TCY Page is happy ✅ 

### Emulator (iOS)

- Tried again and still haven't been able to succesfully run the build according to the README, seeing a *lot* of compile errors.
- Tried deleting the repo and re-cloning it (the `react-native-clean` command nuked my `.git` somehow anyway so I was forced to) and the second time around it *did* work though, maybe worth adding to the testing instructions of this PR so we all start from a clean base.

- Seeing all of my previous native wallets from the previous build around :fridaydog: ✅ 

<img width="496" alt="image" src="https://github.com/user-attachments/assets/95bedca6-8237-4f2a-a464-14d516fcd0a4" />


- Wallet creation is happy ✅ 
- @NeOMakinG Not a mobile app issue per se (would happen on mobile web too), but definitely feels like a bug with the infinite skeleton and empty "My Assets". We probably want an empty state here for assets, and to fix the infinite loading chart, especially since the accounts route isn't exposed on mobile

<img width="488" alt="image" src="https://github.com/user-attachments/assets/d9264de6-aa98-4ead-9402-d87ef9ed352f" />

- Copy is just as sad on simulator as it is on Expo Go 🚫 

- Camera is sad 🚫 (not introduced by this, I did see this the last time around testing simulator)

<img width="523" alt="image" src="https://github.com/user-attachments/assets/433837b2-e42e-433a-889f-e2fcaea17ac9" />

- Onramper is happy ✅ 
- Keystore import is happy ✅ 
- Swaps are happy ✅ 
- TCY is happy (account selection styles borked on mobile, though out of scope of this obviously) ✅ 

<img width="506" alt="image" src="https://github.com/user-attachments/assets/87320787-1440-4cac-a265-d1c313472c18" />

- FOX Page is happy ✅ 
- RFOX is happy ✅ (same styles borked, same unrelated):

<img width="523" alt="image" src="https://github.com/user-attachments/assets/3081c070-8cbd-447a-897d-a32faabcff99" />

@NeOMakinG This PR seems happy for the most part, main two things here before we go with this would be to:

1. fix copy functionality
2. find a way to ensure that camera actually works. i.e if it's borked both on simulator *and* on actual Expo Go on device, how can we confirm it actually still works?

NeOMakinG commented:
gm @gomesalexandre 

Nice catch for the copy feature, fixed it in [b7d3e5c](https://github.com/shapeshift/mobile-app/pull/109/commits/b7d3e5c53b629aaef2df610785d6a58690050b43)

The camera is actually working, you are seeing the dummy camera from your emulator

If you run the app through a real build emulator, you'll see that the refresh is not here anymore, it's an Expo Go thing 👍 

I can confirm it works on both android and ios:
<img width="318" alt="image" src="https://github.com/user-attachments/assets/f9487cb1-481f-48b7-85a8-be930986903b" />
<img width="412" alt="image" src="https://github.com/user-attachments/assets/07e22486-c5c6-4e5e-92fc-60c034059c27" />

Those screens confirm the camera is running in the simulator, if you want to try the camera on expo go you can select a production/ephemeral environment and the camera will be accessible as this is a secure environment :)

Here is an example using expo go with `https://app.shapeshift.com`

![image](https://github.com/user-attachments/assets/ae4fa7ed-cbd2-487e-99b1-a7a1297e965c)
![image](https://github.com/user-attachments/assets/849f61c0-06fb-45b7-aade-1cc6c4b09afc)

Considering your tests and mine, pretty confident on this now 🤟

NeOMakinG commented:
I removed the react-native-clean command as it's not supported plus shouldn't be needed, added the `expo prebuild` command which is supposed to do the same thing but with less risks and taking in account the expo bits, it should achieve the same or even better fyi @gomesalexandre

gomesalexandre reviewed (approved):
Restested locally:

- Swaps ✅ 
- Copy ✅ 
- Camera is happy in Expo with prod env ✅ 
- Onramper is happy with prod env ✅ 
- Smoke test is happy ✅ 

Feeling pretty happy to give this guy a stamp, assuming things are happy for large @kaladinlight on Android too!

kaladinlight commented:
- Expo go working on android :heavy_check_mark: 
- Wallet creation working :heavy_check_mark: 
- Wallet import working :heavy_check_mark: 
- General interaction working (swaps, balance, activity, etc) :heavy_check_mark: 
- Keyboard overlap (I was trying to name my wallet, same with wallet import mnemonic entry) :x: 
![image](https://github.com/user-attachments/assets/aff44bfc-f981-47d9-982a-86d52d5733d8)
- Camera does not open or prompt permissions on scan qr code :x: 
![image](https://github.com/user-attachments/assets/24bb47e2-352a-4ce0-9c2d-4f8cf45a8697)
- App is not respecting view boundaries (overflowing into system notification bar, gesture footer, and lacking horizontal padding)
![image](https://github.com/user-attachments/assets/18ff96ca-e00f-4d01-bc0b-33d66ae25147)

kaladinlight commented:
@NeOMakinG 

- Keyboard overlap :heavy_check_mark: 
- Screen wiggle :heavy_check_mark: (confirmed this behavior was in the old version as well)
- View boundaries :x: (The top and side boundaries look the same to me. I don't know when/why, but I do occassionally see the bottom view boundary respected)
- Camera situation sounds sane to me :heavy_check_mark:

kaladinlight commented:
Ok, I have confirmed the view boundaries are working when pointing at: https://github.com/shapeshift/web/pull/9635, however I am seeing some oddities around app reloads causing the boundaries to reset which requires an expo `r` reload to restore.

Specifically the reproducable scenarios I have are:
- On localhost env verify boundaries are set, navigate to Explore -> TCY, I then get an Oops! page and upon selecting `Try again` the app reloads without the boundaries set.
- On localhost env verify bondaries are set, switch to prod env and notice boundaries are still set correctly, go back to localhost env and notice the boundaries are not longer set correctly.

kaladinlight reviewed (approved):
View boundaries :heavy_check_mark:
